### PR TITLE
[FLINK-34352][doc] Improve the documentation of allowNonRestoredState

### PR DESCRIPTION
**Pull request by masteryhx**
## What is the purpose of the change

Current documentation of allowNonRestoredState is not clear, we should clarify:

- It can lead to serious issues with correctness if it's used incorrectly.
- The correctness is related to the topological order and the logic of job when removing operator by default.
- For DataStream Job, the operator uid could be assigned explicitly to avoid the reassignment of operator uid.

## Brief change log

  - Add Warning of allowNonRestoredState in the doc

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by masteryhx** [comment]
@Zakelly @wuchong @lincoln-lil Could you help to review this ? Thanks a lot!

File: docs/content/docs/ops/state/savepoints.md [hunk 1]
```diff
@@ -208,7 +208,13 @@ This submits a job and specifies a savepoint to resume from. You may give a path
 
 #### Allowing Non-Restored State
 
-By default, the resume operation will try to map all state of the savepoint back to the program you are restoring with. If you dropped an operator, you can allow to skip state that cannot be mapped to the new program via `--allowNonRestoredState` (short: `-n`) option:
+By default, the resume operation will try to map all state of the savepoint back to the program you are restoring with. If you dropped an operator, you can allow to skip state that cannot be mapped to the new program via `--allowNonRestoredState` (short: `-n`) option.
+
+{{< hint warning  >}}
+Improper usage of this feature could result in significant issues with the correctness of the application. It is crucial to verify that any remaining states can be accurately mapped to the appropriate operators.
+It is worth noting that operator UIDs are reassigned based on topological order by default, which may lead to incorrect associations between states and operators.
```
Comment (Zakelly) [inline docs/content/docs/ops/state/savepoints.md:215]: More description for consequence? Such as
```suggestion
It is worth noting that operator UIDs are reassigned based on topological order by default, which may lead to incorrect associations between states and operators, thus consequently the state is not correctly restored as wished.
```

File: docs/content/docs/ops/state/savepoints.md [hunk 2]
```diff
@@ -208,7 +208,13 @@ This submits a job and specifies a savepoint to resume from. You may give a path
...
+By default, the resume operation will try to map all state of the savepoint back to the program you are restoring with. If you dropped an operator, you can allow to skip state that cannot be mapped to the new program via `--allowNonRestoredState` (short: `-n`) option.
+
+{{< hint warning  >}}
+Improper usage of this feature could result in significant issues with the correctness of the application. It is crucial to verify that any remaining states can be accurately mapped to the appropriate operators.
+It is worth noting that operator UIDs are reassigned based on topological order by default, which may lead to incorrect associations between states and operators. 
+To prevent such mismatches, it is advisable to explicitly [assign UIDs] ({{< ref "docs/ops/production_ready" >}}/#set-uuids-for-all-operators) to all operators in a DataStream job. In other scenarios involving mismatches, it is necessary to start the new job discarding all states.
```
Comment (Zakelly) [inline docs/content/docs/ops/state/savepoints.md:216]: what do you mean by 'In other scenarios involving mismatches, it is necessary to start the new job discarding all states.'

**Review by Zakelly — COMMENTED** [review]
Thanks for your contribution! I left couple of comments, PTAL.

File: docs/content/docs/ops/state/savepoints.md (hunk 2 above)
Comment (masteryhx) [reply docs/content/docs/ops/state/savepoints.md]: The description is somewhat confusing and redundant. I just removed it.

**Comment by masteryhx** [comment]
@Zakelly Thanks for the review. I have addressed your comments. PTAL again.

File: docs/content.zh/docs/ops/state/savepoints.md [hunk 3]
```diff
@@ -180,7 +180,15 @@ $ bin/flink run -s :savepointPath [:runArgs]
...
 
-默认情况下，resume 操作将尝试将 Savepoint 的所有状态映射回你要还原的程序。 如果删除了运算符，则可以通过 `--allowNonRestoredState`（short：`-n`）选项跳过无法映射到新程序的状态：
+默认情况下，resume 操作将尝试将 Savepoint 的所有状态映射回你要还原的程序。 如果删除了运算符，则可以通过 `--allowNonRestoredState`（short：`-n`）选项跳过无法映射到新程序的状态。
+
+{{< hint warning  >}}
+错误使用这个功能可能会导致应用程序出现严重的正确性问题。其中最重要的是确认剩余的状态是否能够准确地映射到相应的运算符上。
+
+值得注意的是，默认情况下，操作符的 UID 是根据拓扑顺序重新分配的，这可能会导致状态和运算符之间的关联不正确，最终造成状态未能按照预期正确恢复。
+
+为了防止这种不匹配，建议在 DataStream 作业中显式地为所有操作符[指定 UID ]({{< ref "docs/ops/production_ready" >}}/#set-uuids-for-all-operators)。对其他不匹配的场景，需要舍弃所有状态并启动新作业。
```
Comment (Zakelly) [inline docs/content.zh/docs/ops/state/savepoints.md:190]:
```suggestion
为了防止这种不匹配，建议在 DataStream 作业中显式地为所有运算符[指定 UID ]({{< ref "docs/ops/production_ready" >}}/#set-uuids-for-all-operators)。
```

File: docs/content.zh/docs/ops/state/savepoints.md [hunk 4]
```diff
@@ -180,7 +180,15 @@ $ bin/flink run -s :savepointPath [:runArgs]
 
 #### 跳过无法映射的状态恢复
 
-默认情况下，resume 操作将尝试将 Savepoint 的所有状态映射回你要还原的程序。 如果删除了运算符，则可以通过 `--allowNonRestoredState`（short：`-n`）选项跳过无法映射到新程序的状态：
+默认情况下，resume 操作将尝试将 Savepoint 的所有状态映射回你要还原的程序。 如果删除了运算符，则可以通过 `--allowNonRestoredState`（short：`-n`）选项跳过无法映射到新程序的状态。
+
+{{< hint warning  >}}
+错误使用这个功能可能会导致应用程序出现严重的正确性问题。其中最重要的是确认剩余的状态是否能够准确地映射到相应的运算符上。
+
+值得注意的是，默认情况下，操作符的 UID 是根据拓扑顺序重新分配的，这可能会导致状态和运算符之间的关联不正确，最终造成状态未能按照预期正确恢复。
```
Comment (Zakelly) [inline docs/content.zh/docs/ops/state/savepoints.md:188]:
```suggestion
需要注意在默认情况下，运算符的 UID 是根据拓扑顺序重新分配的，这会导致状态与运算符无法正确关联，最终造成状态未能按照预期正确恢复。
```

**Review by Zakelly — COMMENTED** [review]
Hi @masteryhx , I suggest minor adjustments in Chinese. PTAL, thanks!

**Comment by masteryhx** [comment]
> Hi @masteryhx , I suggest minor adjustments in Chinese. PTAL, thanks!

Thanks for the suggestions, I have addressed and rebased commits. PTAL again.

**Review by Zakelly — APPROVED** [review]
LGTM. Thanks for the update

**Pull request closed without merging**